xla/service/gpu/transforms/cudnn_fused_conv_rewriter.cc

```diff
@@ -438,27 +384,34 @@ class GraphString {
       return false;
     }
 
-    // Insert the op ahead of its first use as an operand of another op.
+    // Insert the op in front of its first use as an operand of another op.
     auto pos = std::find_if(
         graph_.begin(), graph_.end(), [&op](OpDescriptor graph_op) -> bool {
           return std::find(graph_op.operands.begin(), graph_op.operands.end(),
                            op) != graph_op.operands.end();
         });
     pos = graph_.insert(pos, OpDescriptor{op, element_type, op_name, operands});
 
-    // If necessary, move the operands of the op already in the graph ahead of
-    // the op.
-    for (HloInstruction* operand : operands) {
-      auto operand_pos = std::find_if(
-          pos, graph_.end(), [&operand](OpDescriptor graph_op) -> bool {
-            return operand == graph_op.instr;
-          });
-      if (operand_pos != graph_.end()) {
-        OpDescriptor operand_desc = *operand_pos;
-        graph_.erase(operand_pos);
-        pos = graph_.insert(pos, operand_desc);
-      }
-    }
+    // If necessary, move the operands of the op already in the graph in front
+    // of the op. Recursively repeat for the operands' operands.
+    std::function<void(std::vector<HloInstruction*>,
+                       std::vector<OpDescriptor>::iterator)>
+        reorder_operands = [&](auto operands, auto pos) {
+          for (HloInstruction* operand : operands) {
+            auto operand_pos = std::find_if(
+                pos, graph_.end(), [operand](OpDescriptor graph_op) -> bool {
```

Comment:
You look from `pos`, but the effective position of the instruction might change due to recursive calls to `reorder_operands`.

You should sort after the graph is built. You can call `MakeInstructionPostOrder` on the Hlo module, and for each, if its in the cudnn graph, put it in the list. This is short, simple, and only takes O(n) time